README
A documentation how to install a PECL-Extension is available in the PHP-Manual http://www.php.net/manual/en/install.pecl.php

The documentation is now available in the PHP-Manual http://www.php.net/manual/en/
For a more up-to-date documentation refer to the livedocs http://livedocs.phpdoc.info/index.php?l=en&q=ref.gnupg

Note This package is a part of Debian/Ubuntu merge  happened sometime in 2014, apparently It *only* took them *six* years
So well done Debian PHP team, also I love the fact that I am not credited at *all* for the patch :). But what can ya do 


debian/changelog
php-gnupg (1.3.1-0ubuntu1) jaunty; urgency=low

  * Initial Release (LP: #238038).
  * Vcs now Git
  * Change my mail back to what it was 

 -- Matt Arnold <mattarnold5@gmail.com>  Thu, 27 Nov 2008 19:16:19 -0500


php_gnupg.h
/*
  +--------------------------------------------------------------------+
  | PECL :: gnupg                                                      |
  +--------------------------------------------------------------------+
  | Redistribution and use in source and binary forms, with or without |
  | modification, are permitted provided that the conditions mentioned |
  | in the accompanying LICENSE file are met.                          |
  +--------------------------------------------------------------------+
  | Copyright (c) 2006, Thilo Raufeisen <traufeisen@php.net>           |
  +--------------------------------------------------------------------+
*/

/* $Id: php_gnupg.h,v 1.16 2006/11/03 12:38:35 traufeisen Exp $ */

#ifndef PHP_GNUPG_H
#define PHP_GNUPG_H

extern zend_module_entry gnupg_module_entry;
#define phpext_gnupg_ptr &gnupg_module_entry

#ifdef PHP_WIN32
#define PHP_GNUPG_API __declspec(dllexport)
#else
#define PHP_GNUPG_API
#endif

#ifdef ZTS
#include "TSRM.h"
#endif

#include <gpgme.h>

typedef struct gnupg_object{
	zend_object zo;
	gpgme_ctx_t ctx;
    gpgme_error_t err;
    int errormode;
    char* errortxt;
    int signmode;
    gpgme_key_t *encryptkeys;
    unsigned int encrypt_size;
    HashTable *signkeys;
    HashTable *decryptkeys;
} gnupg_object;

zend_class_entry *gnupg_class_entry;

PHP_MINIT_FUNCTION(gnupg);
PHP_MSHUTDOWN_FUNCTION(gnupg);
PHP_MINFO_FUNCTION(gnupg);

PHP_FUNCTION(gnupg_keyinfo);
PHP_FUNCTION(gnupg_verify);
PHP_FUNCTION(gnupg_geterror);
PHP_FUNCTION(gnupg_setsignmode);
PHP_FUNCTION(gnupg_setarmor);
PHP_FUNCTION(gnupg_sign);
PHP_FUNCTION(gnupg_clearsignkeys);
PHP_FUNCTION(gnupg_clearencryptkeys);
PHP_FUNCTION(gnupg_cleardecryptkeys);
PHP_FUNCTION(gnupg_getprotocol);
PHP_FUNCTION(gnupg_encrypt);
PHP_FUNCTION(gnupg_encryptsign);
PHP_FUNCTION(gnupg_decrypt);
PHP_FUNCTION(gnupg_decryptverify);
PHP_FUNCTION(gnupg_export);
PHP_FUNCTION(gnupg_import);
PHP_FUNCTION(gnupg_init);
PHP_FUNCTION(gnupg_addsignkey);
PHP_FUNCTION(gnupg_addencryptkey);
PHP_FUNCTION(gnupg_adddecryptkey);
PHP_FUNCTION(gnupg_deletekey);
PHP_FUNCTION(gnupg_gettrustlist);
PHP_FUNCTION(gnupg_listsignatures);
PHP_FUNCTION(gnupg_seterrormode);

#ifdef ZTS
#define GNUPG_G(v) TSRMG(gnupg_globals_id, zend_gnupg_globals *, v)
#else
#define GNUPG_G(v) (gnupg_globals.v)
#endif

#endif	/* PHP_GNUPG_H */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


php_gnupg_keylistiterator.h
/*
  +--------------------------------------------------------------------+
  | PECL :: gnupg                                                      |
  +--------------------------------------------------------------------+
  | Redistribution and use in source and binary forms, with or without |
  | modification, are permitted provided that the conditions mentioned |
  | in the accompanying LICENSE file are met.                          |
  +--------------------------------------------------------------------+
  | Copyright (c) 2006, Thilo Raufeisen <traufeisen@php.net>           |
  +--------------------------------------------------------------------+
*/

/* $Id: php_gnupg_keylistiterator.h,v 1.6 2006/11/03 12:38:35 traufeisen Exp $ */

#ifndef PHP_GNUPG_KEYLISTITERATOR_H
#define PHP_GNUPG_KEYLISTITERATOR_H

extern zend_module_entry gnupg_keyiterator_module_entry;

#ifdef PHP_WIN32
#define PHP_GNUPG_API __declspec(dllexport)
#else
#define PHP_GNUPG_API
#endif

#ifdef ZTS
#include "TSRM.h"
#endif

#include <gpgme.h>

#define gnupg_keylistiterator_init() _gnupg_keylistiterator_init(INIT_FUNC_ARGS_PASSTHRU)
extern int  _gnupg_keylistiterator_init(INIT_FUNC_ARGS);

typedef struct _gnupg_keylistiterator_object{
	zend_object zo;
	gpgme_ctx_t ctx;
    gpgme_error_t err;
    gpgme_key_t gpgkey;
    zval pattern;
} gnupg_keylistiterator_object;

zend_class_entry *gnupg_keylistiterator_class_entry;

PHP_METHOD(gnupg_keylistiterator, __construct);
PHP_METHOD(gnupg_keylistiterator, current);
PHP_METHOD(gnupg_keylistiterator, next);
PHP_METHOD(gnupg_keylistiterator, rewind);
PHP_METHOD(gnupg_keylistiterator, key);
PHP_METHOD(gnupg_keylistiterator, valid);


#ifdef ZTS
#define GNUPG_G(v) TSRMG(gnupg_globals_id, zend_gnupg_globals *, v)
#else
#define GNUPG_G(v) (gnupg_globals.v)
#endif

#endif	/* PHP_GNUPG_H */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


gnupg_keylistiterator.c
/*
  +--------------------------------------------------------------------+
  | PECL :: gnupg                                                      |
  +--------------------------------------------------------------------+
  | Redistribution and use in source and binary forms, with or without |
  | modification, are permitted provided that the conditions mentioned |
  | in the accompanying LICENSE file are met.                          |
  +--------------------------------------------------------------------+
  | Copyright (c) 2006, Thilo Raufeisen <traufeisen@php.net>           |
  +--------------------------------------------------------------------+
*/

/* $Id: gnupg_keylistiterator.c,v 1.9 2006/11/03 12:38:35 traufeisen Exp $ */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php.h"

#ifdef ZEND_ENGINE_2

#include "php_ini.h"
#include "ext/standard/info.h"
#include "zend_interfaces.h"
#include "php_gnupg.h"
#include "php_gnupg_keylistiterator.h"

static int le_gnupg_keylistiterator;

static zend_object_handlers gnupg_keylistiterator_object_handlers;

/* {{{ defs */
#define GNUPG_GET_ITERATOR() \
    zval *this = getThis(); \
    gnupg_keylistiterator_object *intern; \
    if(this){ \
        intern  =   (gnupg_keylistiterator_object*) zend_object_store_get_object(getThis() TSRMLS_CC); \
        if(!intern){ \
            php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid or unitialized gnupg object"); \
            RETURN_FALSE; \
        } \
    }
/* }}} */

/* {{{ free_iterator_storage */
static void gnupg_keylistiterator_dtor(gnupg_keylistiterator_object *intern TSRMLS_DC){
	if(!intern){
		return;
	}
	gpgme_op_keylist_end(intern->ctx);
	gpgme_key_release(intern->gpgkey);
	gpgme_release(intern->ctx);
/*
	zval_dtor(&intern->pattern);
*/	
	if(intern->zo.properties){
		zend_hash_destroy(intern->zo.properties);
		FREE_HASHTABLE(intern->zo.properties);
	}
	efree(intern);
}
/* }}} */

/* {{{ keylistiterator_objects_new */
zend_object_value gnupg_keylistiterator_objects_new(zend_class_entry *class_type TSRMLS_DC){
	gnupg_keylistiterator_object *intern;
	zend_object_value retval;
	gpgme_ctx_t ctx;

	intern =	emalloc(sizeof(gnupg_keylistiterator_object));
	intern->zo.ce = class_type;
	intern->zo.properties = NULL;
	retval.handle   =   zend_objects_store_put(intern,NULL,(zend_objects_free_object_storage_t) gnupg_keylistiterator_dtor,NULL TSRMLS_CC);
	retval.handlers	=	(zend_object_handlers *) & gnupg_keylistiterator_object_handlers;

    gpgme_new(&ctx);
	intern->ctx		=	ctx;
	return retval;
}
/* }}} */

/* {{{ methodlist gnupg_keylistiterator */
static zend_function_entry gnupg_keylistiterator_methods[] = {
	PHP_ME(gnupg_keylistiterator,	__construct,	NULL,	ZEND_ACC_PUBLIC)
	PHP_ME(gnupg_keylistiterator,	current,		NULL,	ZEND_ACC_PUBLIC)
	PHP_ME(gnupg_keylistiterator,	key,			NULL,	ZEND_ACC_PUBLIC)
	PHP_ME(gnupg_keylistiterator,	next,			NULL,	ZEND_ACC_PUBLIC)
	PHP_ME(gnupg_keylistiterator,	rewind,			NULL,	ZEND_ACC_PUBLIC)
	PHP_ME(gnupg_keylistiterator,	valid,			NULL,	ZEND_ACC_PUBLIC)
    {NULL, NULL, NULL}	
};
/* }}} */

/* {{{ _gnupg_keylistiterator_init
 */
int _gnupg_keylistiterator_init(INIT_FUNC_ARGS)
{
	zend_class_entry ce; 
	
	INIT_CLASS_ENTRY(ce, "gnupg_keylistiterator", gnupg_keylistiterator_methods);
	
	ce.create_object 	=	gnupg_keylistiterator_objects_new;
	gnupg_keylistiterator_class_entry = zend_register_internal_class(&ce TSRMLS_CC);
	memcpy(&gnupg_keylistiterator_object_handlers, zend_get_std_object_handlers(), sizeof(zend_object_handlers));
	le_gnupg_keylistiterator = zend_register_list_destructors_ex(NULL, NULL, "ctx_keylistiterator", module_number);
	
	zend_class_implements   (gnupg_keylistiterator_class_entry TSRMLS_CC, 1, zend_ce_iterator);
	
	return SUCCESS;
}
/* }}} */


PHP_METHOD(gnupg_keylistiterator,__construct){
	zval *pattern;

	int args = ZEND_NUM_ARGS();

	GNUPG_GET_ITERATOR();

	if(args > 0){
		if (zend_parse_parameters(args TSRMLS_CC, "|z", &pattern) == FAILURE){
			return;
		}
		intern->pattern = *pattern;
		zval_copy_ctor(&intern->pattern);
	}else{
		ZVAL_EMPTY_STRING(&intern->pattern);
	}
}
PHP_METHOD(gnupg_keylistiterator,current){
	GNUPG_GET_ITERATOR();
	
	RETURN_STRING(intern->gpgkey->uids[0].uid,1);
}

PHP_METHOD(gnupg_keylistiterator,key){
    GNUPG_GET_ITERATOR();
	
	RETURN_STRING(intern->gpgkey->subkeys[0].fpr,1);
}

PHP_METHOD(gnupg_keylistiterator,next){
	GNUPG_GET_ITERATOR();

	if(intern->gpgkey){
		gpgme_key_release(intern->gpgkey);
	}

	if(intern->err = gpgme_op_keylist_next(intern->ctx, &intern->gpgkey)){
		gpgme_key_release(intern->gpgkey);
		intern->gpgkey = NULL;
	}
	RETURN_TRUE;
}

PHP_METHOD(gnupg_keylistiterator,rewind){
	GNUPG_GET_ITERATOR();

	if((intern->err = gpgme_op_keylist_start(intern->ctx, Z_STRVAL(intern->pattern), 0)) != GPG_ERR_NO_ERROR){
		zend_throw_exception(zend_exception_get_default(),gpg_strerror(intern->err),1 TSRMLS_CC);
	}
	if((intern->err = gpgme_op_keylist_next(intern->ctx, &intern->gpgkey))!=GPG_ERR_NO_ERROR){
		RETURN_FALSE;
	}
	RETURN_TRUE;
}

PHP_METHOD(gnupg_keylistiterator,valid){
	GNUPG_GET_ITERATOR();

	if(intern->gpgkey!=NULL){
		RETURN_TRUE;
	}else{
		RETURN_FALSE;
	}
}
#endif /* ZEND_ENGINE_2 */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
